### PR TITLE
Hide our misuses of contexts from the public interface

### PR DESCRIPTION
**Pull request by corhere**
Our `context` package is a big pile of sadness which breaks most of the context usage rules and encourages bad practices. And tragically our bad habits around contexts have leaked into a couple of our extension-point interfaces: `auth.AccessController` and `storage.StorageDriver`. Modify the interfaces such that implementations are no longer required to abuse contexts, and unexport the pile of sadness.

### CODE REVIEW

**Comment by milosgajdos** [comment]
@corhere this needs a rebase now

**Comment by milosgajdos** [comment]
Sorry, @corhere but this needs another rebase 🙃 I am aiming to review it today

File: internal/requestutil/util.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"net/http"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+func parseIP(ipStr string) net.IP {
+	ip := net.ParseIP(ipStr)
+	if ip == nil {
+		log.Warnf("invalid remote IP address: %q", ipStr)
```
Comment (milosgajdos) [inline internal/requestutil/util.go:14]: question: I've sometimes wondered why we aren't using the `dcontext.Logger` but instead use the "raw" `logrus`

File: registry/auth/auth.go [hunk 2]
```diff
@@ -93,80 +88,21 @@ type Challenge interface {
 // and required access levels for a request. Implementations can support both
 // complete denial and http authorization challenges.
 type AccessController interface {
-	// Authorized returns a non-nil error if the context is granted access and
+	// Authorized returns a nil error if the request is granted access and
 	// returns a new authorized context. If one or more Access structs are
```
Comment (milosgajdos) [inline registry/auth/auth.go:92]: > returns a new authorized context

this is no longer true 🤔 `Authorized` now returns `Grant`

File: registry/auth/token/token_test.go [hunk 3]
```diff
@@ -477,8 +475,8 @@ func TestAccessController(t *testing.T) {
 		t.Fatalf("accessControler did not get expected error - got %s - expected %s", challenge, ErrTokenRequired)
 	}
 
-	if authCtx != nil {
-		t.Fatalf("expected nil auth context but got %s", authCtx)
+	if grant != nil {
+		t.Fatalf("expected nil auth grant but got %s", grant)
```
Comment (milosgajdos) [inline registry/auth/token/token_test.go:479]: We need to change the formatter string here, though I wonder if it's worth implementing `Stringer` for `Grant` 🤔

File: registry/auth/token/token_test.go [hunk 4]
```diff
@@ -512,8 +510,8 @@ func TestAccessController(t *testing.T) {
...
```
Comment (milosgajdos) [inline registry/auth/token/token_test.go:514]: Same comment as above (there are a few more occurrences of this in this file)

File: registry/storage/driver/gcs/gcs.go [hunk 5]
```diff
@@ -810,40 +810,24 @@ func storageCopyObject(ctx context.Context, srcBucket, srcName string, destBucke
...
+		return "", nil
 	}
 
 	opts := &storage.SignedURLOptions{
 		GoogleAccessID: d.email,
 		PrivateKey:     d.privateKey,
-		Method:         methodString,
-		Expires:        expiresTime,
+		Method:         r.Method,
+		Expires:        time.Now().Add(20 * time.Minute),
```
Comment (milosgajdos) [inline registry/storage/driver/gcs/gcs.go:828]: ignore this comment: We really need to make this (expiry) a config option

File: registry/storage/driver/storagedriver.go [hunk 6]
```diff
@@ -92,11 +93,10 @@ type StorageDriver interface {
...
 
-	// URLFor returns a URL which may be used to retrieve the content stored at
-	// the given path, possibly using the given options.
-	// May return an ErrUnsupportedMethod in certain StorageDriver
-	// implementations.
-	URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error)
+	// RedirectContent returns a URL which the client of the request r may use
+	// to retrieve the content stored at path. Returning the empty string
+	// signals that the request may not be redirected.
+	RedirectContent(r *http.Request, path string) (string, error)
```
Comment (milosgajdos) [inline registry/storage/driver/storagedriver.go:99]: nit: I think I'd prefer this to be called`ContentURL` so it's more obvious from the name of the func what's returned (yes you can always check godoc)

**Review by milosgajdos — COMMENTED** [review]
This is so nice @corhere. A long belated cleanup 🧹 Reading some of that code felt like walking through `Go` purgatory 😄 

There is a couple of comments I left that we should address (godoc comment and code formatter in some places) but otherwise LGTM

File: internal/dcontext/doc.go [hunk 7]
```diff
@@ -27,7 +27,7 @@
 // the context and reported with the logger. The following example would
 // return a logger that prints the version with each log message:
 //
-//	ctx := context.Context(context.Background(), "version", version)
+//	ctx := context.WithValue(dcontext.Background(), "version", version)
```
[the remark this answers is not in the document]
Comment (corhere) [reply internal/dcontext/doc.go:30]: No function exists with that name. There is `dcontext.WithValues` (note the plural) which takes a `map[string]any` argument and returns a context which is functionally indistinguishable from a context produced from a chain of `context.WithValue` calls.

[the remark this answers is not in the document]
Comment (milosgajdos) [reply internal/dcontext/doc.go:30]: Ah, I missed the `s` 😬

File: registry/storage/driver/storagedriver.go (hunk 6 above)
Comment (corhere) [reply registry/storage/driver/storagedriver.go:99]: The function's semantics are more (subtle) than just what is returned. It is not simply returning "the" URL for the content. The driver is dictating _whether_ a particular request for content should be redirected, and if so, where to redirect that request to. The s3 and gcs drivers return a different URL depending on the request's HTTP verb. The cloudfront storage middleware decides whether or not to redirect the request by examining the request's source IP.

I had picked the name `RedirectContent` to draw parallels to `GetContent` and `PutContent`, though now looking at it with fresh eyes I realize that the comparison is more likely to confuse than inform readers. What do you think of the name `RedirectURL`?

Comment (milosgajdos) [reply registry/storage/driver/storagedriver.go:99]: `RedirectURL` works for me 👍

File: registry/auth/token/token_test.go (hunk 3 above)
Comment (corhere) [reply registry/auth/token/token_test.go:479]: Good catch! `%s` would not have been terribly effective at formatting a `context.Context` legibly, either.

File: internal/requestutil/util.go (hunk 1 above)
Comment (corhere) [reply internal/requestutil/util.go:14]: I went to go plumb contexts through (from `(*http.Request).Context()`) but realized that requestutil importing dcontext would be an import cycle as dcontext already imports requestutil. Frankly the requestutil functions are not amazing (parse the IP address only to throw away the parsed form and return the source string representation?!) so I think that the the import cycle would best be broken by redesigning the requestutil functions such that the caller can do the logging, if it chooses to.

**Comment by corhere** [comment]
I've pushed changes as fixup commits for your reviewing convenience. Don't worry about the failed DCO check: I'll [autosquash](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---autosquash) the offending fixup commits away after review.

**Comment by milosgajdos** [comment]
PTAL: @thaJeztah @Jamstah @squizzi

**Review by milosgajdos — APPROVED** [review]
LGTM. Once others review we should squash and sign and we're good to go 😄

File: registry/storage/driver/filesystem/driver.go [hunk 8]
```diff
@@ -282,10 +283,9 @@ func (d *driver) Delete(ctx context.Context, subPath string) error {
...
 }
 
-// URLFor returns a URL which may be used to retrieve the content stored at the given path.
-// May return an UnsupportedMethodErr in certain StorageDriver implementations.
-func (d *driver) URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error) {
-	return "", storagedriver.ErrUnsupportedMethod{}
+// RedirectURL returns a URL which may be used to retrieve the content stored at the given path.
+func (d *driver) RedirectURL(*http.Request, string) (string, error) {
+	return "", nil
 }
```
Comment (thaJeztah) [inline registry/storage/driver/filesystem/driver.go:289]: Related to the above; could an alternative be to move `RedirectURL` to a separate interface, which would be an optional feature to be implemented by drivers.

Drivers not needing this feature could then omit implementing it.

File: registry/storage/driver/gcs/gcs.go (hunk 5 above)
Comment (thaJeztah) [reply registry/storage/driver/gcs/gcs.go:828]: ^^ @milosgajdos I know you mention "ignore this comment", but isn't that what's being removed? (but possibly not used currently)? Or do you mean "make it configurable, but _properly_ this time"?

**Review by thaJeztah — APPROVED** [review]
LGTM, nice! 

Left some comments; don't think they're blockers, but perhaps you want to look if any of them make sense and you prefer changing.

File: registry/storage/blobserver.go [hunk 9]
```diff
@@ -35,19 +35,16 @@ func (bs *blobServer) ServeBlob(ctx context.Context, w http.ResponseWriter, r *h
...
-			return err
-
-		case driver.ErrUnsupportedMethod:
-			// Fallback to serving the content directly.
-		default:
-			// Some unexpected error.
-			return err
+			return nil
 		}
+		// Fallback to serving the content directly.
```
[the remark this answers is not in the document]
Comment (corhere) [reply registry/storage/blobserver.go:47]: > would there be any possible situations where an empty string is returned (but no error), and we fall-through; and are there any possible consequences to falling through?

Yes, and yes: the blobserver holds up its end of the interface contract.

```go
	// RedirectURL returns a URL which the client of the request r may use
	// to retrieve the content stored at path. Returning the empty string
	// signals that the request may not be redirected.
```

File: registry/storage/driver/filesystem/driver.go (hunk 8 above)
Comment (corhere) [reply registry/storage/driver/filesystem/driver.go:289]: I had considered that, but [it's really awkward and verbose](https://blog.merovius.de/posts/2017-07-30-the-trouble-with-optional-interfaces/#they-remove-a-lot-of-the-power-of-interfaces) to make a wrapper struct type (read: storage middleware) which optionally implements an interface iff the inner type implements the same interface. Instead I tweaked the method's contract to make it less weird that it's mandatory.

**Pull request merged**